README.md
# graylog2

[![NPM version](http://img.shields.io/npm/v/graylog2.svg?style=flat-square)](https://www.npmjs.org/package/graylog2)
[![NPM license](http://img.shields.io/npm/l/graylog2.svg?style=flat-square)](https://www.npmjs.org/package/graylog2)

Graylog2 client library for Node.js

## Installation

```sh
npm install graylog2 --save
```

## What is Graylog?

Graylog is popular logging software. You can get it at http://www.graylog2.org. Incidentally, since this package
uses Graylog's [GELF](http://docs.graylog.org/en/latest/pages/gelf.html) protocol, you can also use this with other
logging software that has GELF support.

## Usage

### Available functions

* graylog.emergency(short, full, fields, timestamp)
* graylog.alert(short, full, fields, timestamp)
* graylog.critical(short, full, fields, timestamp)
* graylog.error(short, full, fields, timestamp)
* graylog.warning(short, full, fields, timestamp)
* graylog.notice(short, full, fields, timestamp)
* graylog.info(short, full, fields, timestamp)
* graylog.debug(short, full, fields, timestamp)

Arguments:

- short (string): A short message to log.
- full (string, optional): Additional details.
- fields (object, optional): An object of key/value pairs to help with filtering.
- timestamp (integer, optional): A custom timestamp (milliseconds).

### Code snippets

```js
var Graylog = require('graylog2');
var logger = new Graylog({
    servers: [
        { host: '127.0.0.1', port: 12201 },
        { host: '127.0.0.2', port: 12201 }
    ],
    hostname: 'server.name', // the name of this host (optional, default: os.hostname())
    facility: 'Node.js',     // the facility for these log messages (optional, default: "Node.js")
    bufferSize: 1350         // max UDP packet size, should not exceed the MTU of your network (optional, default: 1400)
});

logger.on('error', function (error) {
    console.error('Error while trying to write to Graylog2:', error);
});

logger.on('warning', function (error) {
	console.error('Non-fatal error while trying to write to Graylog2:', error);
});
```

Short message:

```js
logger.debug("What we've got here is...failure to communicate");
```

Long message:

```js
var short = "What we've got here is...failure to communicate";
var long = "Some men you just can't reach. So you get what we had here last week, " +
	"which is the way he wants it... well, he gets it. I don't like it any more than you men.";

logger.debug(short, long);
```

Short with additional data:

```js
logger.debug("What we've got here is...failure to communicate", { cool: 'beans' });
```

Long with additional data:

```js
var short = "What we've got here is...failure to communicate";
var long = "Some men you just can't reach. So you get what we had here last week, " +
	"which is the way he wants it... well, he gets it. I don't like it any more than you men.";

logger.debug(short, long, { cool: 'beans' });
```

Send all pending log messages and close the socket:

```js
logger.close(function () {
    console.log('All done!');
});
```

### More examples

    npm install graylog2

## Graylog2 Configuration

This module will send its data as GELF packets to Graylog2. In order to see your data in the correct format you need to create a GELF Input in your Graylog2 application. 

You can do this by following these instructions:

1. Go to System -> Inputs

<div align="center">
    <img src="./imgs/graylog_config_1.png">
</div>

2. Select a GELF Input type. In this case we will be using GELF UDP as it doesn't need any additional configuration.

<div align="center">
    <img src="./imgs/graylog_config_2.png">
</div>

3. Select the Nodes that will read the new Input type.

<div align="center">
    <img src="./imgs/graylog_config_3.png">
</div>

4. Launch the new input!

<div align="center">
    <img src="./imgs/graylog_config_4.png">
</div>


CHANGELOG
0.3.0: June 29, 2018
* Updated README.md with graylog configuration
* Fixed several vulnerability issues with dependencies
* Fixed security issue with usage of Buffer
* Fixed https://github.com/Wizcorp/node-graylog2/issues/16

0.2.1: February 5, 2018
* Fixed a typo in error file/line parsing.

0.2.0: September 20, 2017
* Bumped the GELF spec version to 1.1.

0.1.3: July 16, 2014
* Added asynchronous close(cb) function that closes and returns when all messages in the queue have been sent.

0.1.1: October 29, 2013
* The Emergency level was yielding "undefined", making Graylog2 fall back to the default Alert level.
* According to documentation log method should handle following invocation:
  `client.log('short message', {param: "value"})`, but instead of sending additional fields, JSON
  was treated like full message field.

0.1.0: September 3, 2013
* Replaced all console.error calls with emit('error').
  *Warning*: you really must listen for these or node will throw uncatchable exceptions!
* Bugfix: chunk count calculation was ignoring the header-overhead.
* All log operations now use a single udp socket.
* To be more friendly to lower MTUs, lowered the default buffer size to 1400 bytes.
* Added destroy() function that closes the UDP socket.
* Removed async requirement.
* Removed some unused variables.
* Removed the larger test files (256 and 512KB) because they don't fit the new default buffer limit.
* Removed unused test data.

0.0.12: March 12, 2012
* Dummy version bump in order to push to npm

0.0.11: March 12, 2012
* Removed dependency on node-compress-buffer
* Code cleanup
* Contributors mentioned

0.0.10: November 6, 2011
* Support node 0.6.0+

0.0.9: September 19, 2011
* Work around node 0.5.x crashes
* Added 0.5.x notes
* Added an important requirement of compress-buffer >= 0.4.0
* Added sayHello.js

0.0.8: July 12, 2011
* Minor console log output issue

0.0.7: June 23, 2011
* So much better console output

0.0.6: June 23, 2011
* GLOBAL.graylogSequence

0.0.5: May 14, 2011
* GLOBAL.graylogFacility
* Limit message size to 8k
* Misc typo fixes

0.0.2, 0.0.3, 0.0.4: May 12, 2011
* Exceptions in case the graylog2 is not running
* GLOBAL.graylogToConsole added
* CHANGELOG added
* Bugfixes

0.0.1: May 11, 2011
* initial release


test.js
var graylog = require('./graylog'),
    fs = require('fs'),
    file,
    data,
    servers = [
        { 'host': '127.0.0.1', 'port': 12201 }
    ];


var client = new graylog.graylog({
        servers: servers,
        facility: 'Test logger / Node.JS Test Script'
    });

console.log('---------------------------------------------');
console.log('Sending three test as info, warning and error');
console.log('---------------------------------------------');
client.log('test1', 'i get this1', {cool: 'beans'});
client.warn('test2', 'i get this2', {cool: 'beans'});
client.error('test3', 'i get this3', {cool: 'beans'});
client.error('customTime', 'i get this3', {cool: 'beans'}, new Date('2012-10-10 13:20:31.619Z'));
console.log('');

console.log('---------------------------------------------');
console.log('Sending Sean Connery\' picture (as critical)');
console.log('---------------------------------------------');
file = './data/sean.jpg';
data = fs.readFileSync(file);
client.critical('My Nice Sean Connery Picture', data.toString(), {name: 'James Bond'});
console.log('');

console.log('---------------------------------------------');
console.log('Sending data of different sizes (as critical)');
console.log('---------------------------------------------');
for (var i = 4; i <= 128; i *= 2) {
    file = './data/' + i + '.dat';
    data = fs.readFileSync(file);
    console.log('sending', file);
    client.critical('Test 4 ' + file, data.toString(), {datafile: i + '.dat'});
}
console.log('');

console.log('---------------------------------------------');
console.log('Sending different parameters');
console.log('---------------------------------------------');
client.log('ParametersTest - Only short message');
client.log('ParametersTest - Short message and json', {cool: 'beans'});
client.log('ParametersTest - Short message and full message', 'Full message');
client.log('ParametersTest - Short Message with full message and json', 'Full message', {cool: 'beans'});
console.log('');


console.log('Insertion complete. Please check', 'http://' + servers[0].host + ':3000', 'and verify that insertion was successfull');
console.log('');


test/close.js
var test = require('tape');
var Graylog = require('..');

test('drain', function (t) {
	var client = new Graylog({ servers: [{ host: '127.0.0.1', port: 12345 }], deflate: 'never' });
	var total = 100;
	var sent = 0;
	var received = 0;

	client.on('chunk', function (header, data, i, count) {
		if (i === count - 1) {
			received += 1;
		}
	});

	client.on('drain', function () {
		t.equal(sent, total);
		t.equal(received, total);
		t.end();
	});

	for (var i = 0; i < total; i++) {
		client.info('short message', 'full message'.repeat(1000));
		sent += 1;
	}
});


test('graceful close', function (t) {
	var client = new Graylog({ servers: [{ host: '127.0.0.1', port: 12345 }], deflate: 'never' });
	var total = 100;
	var sent = 0;
	var received = 0;

	client.on('chunk', function (header, data, i, count) {
		if (i === count - 1) {
			received += 1;
		}
	});

	for (var i = 0; i < total; i++) {
		client.info('short message', 'full message'.repeat(1000));
		sent += 1;
	}

	client.close(function () {
		t.equal(sent, total);
		t.equal(received, total);
		t.end();
	});
});


test('instant close', function (t) {
	var client = new Graylog({ servers: [{ host: '127.0.0.1', port: 12345 }] });

	client.close(function () {
		t.end();
	});
});


test('send after close', function (t) {
	var client = new Graylog({ servers: [{ host: '127.0.0.1', port: 12345 }] });

	client.on('warning', function () {
		t.end();
	});

	client.close(function () {
		client.info('short message');
	});
});


test/small.js
var test = require('tape');
var Graylog = require('..');

test('small', function (t) {
	var client = new Graylog({ servers: [{ host: '127.0.0.1', port: 12345 }], hostname: 'foo', facility: 'bar' });

	client.on('message', function (message) {
		message = JSON.parse(message);

		t.equal(message.version, '1.0');
		t.equal(typeof message.timestamp, 'number');
		t.equal(message.host, 'foo');
		t.equal(message.facility, 'bar');
		t.equal(message.level, client.level.INFO)
		t.equal(message.short_message, 'short message');
		t.equal(message.full_message, 'full message');
		t.equal(message.__id, 1);
		t.equal(message._foo, 'bar');
		t.end();
	});

	client.info('short message', 'full message', { id: 1, foo: 'bar' });
});


test('multiserver', function (t) {
	var servers = [
		{ host: '127.0.0.1', port: 12345 },
		{ host: '127.0.0.1', port: 12345 }
	];

	var client = new Graylog({ servers: servers });
	var cursor = 0;
	var count = 0;
	var total = 5;

	client.on('message', function (message, server) {
		t.equal(server, servers[cursor]);

		cursor += 1;
		count += 1;

		if (cursor >= servers.length) {
			cursor = 0;
		}

		if (count === total) {
			t.end();
		}
	});

	for (var i = 0; i < total; i += 1) {
		client.info('short message');
	}
});


test('bad host', function (t) {
	var client = new Graylog({ servers: [{ host: 'foobar', port: 12345 }] });

	client.on('error', function (error) {
		t.equal(error.code, 'ENOTFOUND');
		t.end();
	});

	client.info('short message');
});


bench.js
var Graylog = require('.');
var client;
var servers = [
    { host: '127.0.0.1', port: 12201 }
];

function createClient() {
	client = new Graylog({
	    servers: servers,
	    facility: 'node-graylog benchmark'
	});

	client.on('error', function (error) {
		throw error;
	});
}


var i = 0;
var count = 20000;
var small = 'h'.repeat(1000);
var big = 'h'.repeat(25000);
var bigRandom = require('crypto').randomBytes(20000).toString('base64');

console.log('');

function log(str, label, i, n, cb) {
    if (i === 0) {
        createClient();
        console.time(label + ' x' + n);

        client.on('drain', function () {
            console.timeEnd(label + ' x' + n);

            console.log('Sent:', client.sent, '- Compressed:', client.compressed);
            console.log('');

            if (client.sent !== n) {
                throw new Error('Should have sent: ' + n);
            }

            cb();
        });
    }

    if (i < n) {
        client.log('test', str);
        process.nextTick(log, str, label, i + 1, n, cb);
    }
}

function testSmall(cb) {
    log(small, 'small', 0, 10000, cb);
}

function testBig(cb) {
    log(big, 'big', 0, 5000, cb);
}

function testBigAndRandom(cb) {
    log(bigRandom, 'bigAndRandom', 0, 2000, cb);
}

function end() {
    console.log('Complete.');
    console.log('Please check your logging service and verify that insertion was successful.');
    console.log('');
}

testSmall(function () {
    testBig(function () {
        testBigAndRandom(end);
    });
});


graylog.js
var zlib = require('zlib');
var crypto = require('crypto');
var dgram = require('dgram');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var assert = require('assert');
var MAX_SAFE_INT = 9007199254740991;  // Number.MAX_SAFE_INTEGER


/**
 * Graylog instances emit errors. That means you really really should listen for them,
 * or accept uncaught exceptions (node throws if you don't listen for "error").
 */

function Graylog(config) {
    EventEmitter.call(this);

    this.config = config;

    this.servers = config.servers;
    this.client = null;
    this.hostname = config.hostname || require('os').hostname();
    this.facility = config.facility || 'Node.js';
    this.deflate = config.deflate || 'optimal';
    assert(
      this.deflate === 'optimal' || this.deflate === 'always' || this.deflate === 'never',
      'deflate must be one of "optimal", "always", or "never". was "' + this.deflate + '"');

    this._bufferSize = config.bufferSize || this.DEFAULT_BUFFERSIZE;

    // state
    this._serverIterator = 0;
    this._discard = false;
    this._isSending = false;
    this._firstMessage = null;
    this._lastMessage = null;
	this._headerPool = [];
}

util.inherits(Graylog, EventEmitter);

Graylog.prototype.DEFAULT_BUFFERSIZE = 1400;  // a bit less than a typical MTU of 1500 to be on the safe side

Graylog.prototype.level = {
    EMERG: 0,    // system is unusable
    ALERT: 1,    // action must be taken immediately
    CRIT: 2,     // critical conditions
    ERROR: 3,    // error conditions
    WARNING: 4,  // warning conditions
    NOTICE: 5,   // normal, but significant, condition
    INFO: 6,     // informational message
    DEBUG: 7     // debug level message
};


Graylog.prototype.getServer = function () {
    if (this.servers.length === 1) {
        // common case
        return this.servers[0];
    }

    this._serverIterator += 1;

    if (this._serverIterator >= MAX_SAFE_INT) {
        this._serverIterator = 0;
    }

    return this.servers[this._serverIterator % this.servers.length];
};


Graylog.prototype.getClient = function () {
    if (!this.client) {
        this.client = dgram.createSocket('udp4');

        var that = this;

        this.client.on('error', function (error) {
            that.emit('error', error);
        });
    }

    return this.client;
};


Graylog.prototype.destroy = function () {
	this._discard = true;

    if (this.client) {
        this.client.close();
        this.client.removeAllListeners();
        this.client = null;
		this._firstMessage = null;
		this._lastMessage = null;
		this._headerPool = [];
    }
};

Graylog.prototype.emergency = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.EMERG);
};

Graylog.prototype.alert = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.ALERT);
};

Graylog.prototype.critical = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.CRIT);
};

Graylog.prototype.error = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.ERROR);
};

Graylog.prototype.warning = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.WARNING);
};
Graylog.prototype.warn = Graylog.prototype.warning;

Graylog.prototype.notice = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.NOTICE);
};

Graylog.prototype.info = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.INFO);
};

Graylog.prototype.log = Graylog.prototype.info;

Graylog.prototype.debug = function (short, full, fields, timestamp) {
    return this._log(short, full, fields, timestamp, this.level.DEBUG);
};


function serialize(hostname, facility, short, full, fields, timestamp, level) {
    var message = {
        version: '1.0',
        timestamp: parseInt((timestamp || Date.now()) / 1000, 10),
        host: hostname,
        facility: facility,
        level: level,
        short_message: short === null ? undefined : short,
        full_message: full === null ? undefined : full
    };

    // We insert additional fields

	if (fields) {
		for (var field in fields) {
	        if (field === 'id') {
	            // http://docs.graylog.org/en/2.1/pages/gelf.html
	            message.__id = fields.id;
	        } else {
	            message['_' + field] = fields[field];
	        }
	    }
	}

    return Buffer.from( JSON.stringify(message), 'utf8' );
}


Graylog.prototype._compressMessage = function (msg, cb) {
    if ((msg.buff.length <= this._bufferSize && this.deflate === 'optimal') || this.deflate === 'never') {
        return cb();
    }

	var that = this;

    zlib.deflate(msg.buff, function (error, compressed) {
        if (error) {
            that.emit('warning', error);
            return cb();
        }

        if (compressed.length < msg.buff.length || that.deflate === 'always') {
			msg.buff = compressed;
        }

        return cb();
    });
};


Graylog.prototype._getHeadersFromPool = function (n) {
	for (var i = this._headerPool.length; i < n; i += 1) {
		var header = this._headerPool[i] = Buffer.alloc(12);

		// Set the magic number (bytes 0 and 1)
        header[0] = 30;
        header[1] = 15;

		// Set the chunk sequence number (byte 10)
        header[10] = i;
	}

	return this._headerPool;
};


Graylog.prototype._sendChunked = function (id, message, cb) {
	var maxDataSize = this._bufferSize - 12; // the message part of each chunk is the buffer size - header size
	var chunkCount = Math.ceil(message.length / maxDataSize);

	if (chunkCount > 128) {
		return cb(new Error('Graylog2 message too long: ' + message.length + ' bytes'));
	}

	var client = this.getClient();
	var server = this.getServer();

	var headers = this._getHeadersFromPool(chunkCount);
	var msgOffset = 0;

	for (var i = 0; i < chunkCount; i += 1) {
		var header = headers[i];

		// Set the message id (bytes 2-9)
		id.copy(header, 2);

		// Set the total number of chunks (byte 11)
		header[11] = chunkCount;

		// Slice out the message part
		var data = message.slice(msgOffset, msgOffset + maxDataSize);

		if (i < chunkCount - 1) {
			client.send([header, data], server.port, server.host);

			msgOffset += maxDataSize;
		} else {
			client.send([header, data], server.port, server.host, cb);
		}
	}
};


Graylog.prototype._sendMessage = function (msg, cb) {
	var message = msg.buff;
	msg.buff = null; // help GC a bit

    if (message.length <= this._bufferSize) {
		// No need to chunk this message

		var client = this.getClient();
		var server = this.getServer();

		client.send(message, 0, message.length, server.port, server.host, cb);
		return;
    }

    var that = this;

    // Generate a random ID as a buffer
    crypto.randomBytes(8, function (error, id) {
        if (error) {
            return cb(error);
        }

		that._sendChunked(id, message, cb);
    });
};


Graylog.prototype._send = function () {
    if (this._isSending) {
        // already sending
        return true;
    }

    if (!this._firstMessage) {
        // nothing to send
        this.emit('drain');
        return false;
    }

    this._isSending = true;

    // pull off a message

    var msg = this._firstMessage;

	if (msg.next) {
		this._firstMessage = msg.next;
	} else {
		this._firstMessage = this._lastMessage = null;
	}

    // send the message

    var that = this;

    function onSend(error) {
		that._isSending = false;

        if (error) {
            that.emit('error', error);
            return;
        }

		// because onSend is always called asynchronously, calling _send() immediately is safe
		that._send();
    }

    function onCompress(error) {
        if (error) {
			that._isSending = false;
            that.emit('error', error);
            return;
        }

    	that._sendMessage(msg, onSend);
    }

	this._compressMessage(msg, onCompress);

    return true;
};


Graylog.prototype._log = function log(short, full, fields, timestamp, level) {
    if (this._discard) {
        return;
    }

    var message = {
		buff: serialize(this.hostname, this.facility, short, full, fields, timestamp, level),
		next: null
	};

	// append to queue

    if (this._lastMessage) {
        this._lastMessage.next = message;
		this._lastMessage = message;
    } else {
        this._firstMessage = this._lastMessage = message;
    }

    return this._send();
};


Graylog.prototype.close = function (cb) {
	if (!cb) {
		cb = function () {};
	}

    if (!this._isSending) {
        this.destroy();
        return cb();
    }

    var that = this;

    this.once('drain', function () {
        that.close(cb);
    });
};

exports.graylog = Graylog; // deprecated
exports.Graylog = Graylog;


lib/Graylog.js
var zlib = require('zlib');
var crypto = require('crypto');
var dgram = require('dgram');
var util = require('util');
var EventEmitter = require('events').EventEmitter;
var Queue = require('./Queue');

var MAX_SAFE_INT = 9007199254740991;  // Number.MAX_SAFE_INTEGER

/**
 * Graylog instances emit errors. That means you really really should listen for them,
 * or accept uncaught exceptions (node throws if you don't listen for "error").
 */

function Graylog(config) {
    EventEmitter.call(this);

	// settings (all safe to change at runtime through these setters)
	this.setServers(config.servers);
	this.setHostname(config.hostname || require('os').hostname());
	this.setFacility(config.facility || 'Node.js');
	this.setBufferSize(config.bufferSize || this.DEFAULT_BUFFERSIZE);
	this.setDeflate(config.deflate || 'optimal');

    // state
	this._client = null;
    this._serverIterator = 0;
	this._headerPool = [];
	this._isDeflating = false;
	this._isSending = false;

	this._sendQueue = new Queue();
	this._deflateQueue = new Queue();

	// stats
	this.sent = 0;
	this.compressed = 0;
}

util.inherits(Graylog, EventEmitter);

Graylog.prototype.DEFAULT_BUFFERSIZE = 1400;  // a bit less than a typical MTU of 1500 to be on the safe side

Graylog.prototype.level = {
    EMERG: 0,    // system is unusable
    ALERT: 1,    // action must be taken immediately
    CRIT: 2,     // critical conditions
    ERROR: 3,    // error conditions
    WARNING: 4,  // warning conditions
    NOTICE: 5,   // normal, but significant, condition
    INFO: 6,     // informational message
    DEBUG: 7     // debug level message
};


Graylog.prototype.setServers = function (servers) {
	if (!Array.isArray(servers)) {
		throw new TypeError('Servers must be an array');
	}

	if (servers.length === 0) {
		throw new Error('Servers array cannot be empty');
	}

	for (var i = 0; i < servers.length; i += 1) {
		var server = servers[i];
		if (!server || typeof server !== 'object') {
			throw new TypeError('A server entry must be an object with "host" and "port" properties');
		}

		if (!server.hasOwnProperty('host') || !server.hasOwnProperty('port')) {
			throw new TypeError('A server entry must be an object with "host" and "port" properties');
		}

		if (typeof server.host !== 'string') {
			throw new TypeError('A server host must be a string');
		}

		if (typeof server.port !== 'number') {
			throw new TypeError('A server port must be a number');
		}
	}

	this._servers = servers;
};


Graylog.prototype.setHostname = function (hostname) {
	if (typeof hostname !== 'string') {
		throw new TypeError('Host name must be a string');
	}

	this._hostname = hostname;
};


Graylog.prototype.setFacility = function (facility) {
	if (typeof facility !== 'string') {
		throw new TypeError('Facility must be a string');
	}

	this._facility = facility;
};


Graylog.prototype.setBufferSize = function (bufferSize) {
	if (typeof bufferSize !== 'number') {
		throw new TypeError('Buffer size must be a number');
	}

	this._bufferSize = bufferSize;
};


Graylog.prototype.setDeflate = function (deflate) {
	if (deflate !== 'optimal' && deflate !== 'always' && deflate !== 'never') {
		throw new Error('deflate must be "optimal", "always", or "never". was "' + deflate + '"');
	}

	this._neverDeflate = deflate === 'never';
	this._alwaysDeflate = deflate === 'always';
};


Graylog.prototype._getServer = function () {
    if (this._servers.length === 1) {
        // common case
        return this._servers[0];
    }

    if (this._serverIterator >= MAX_SAFE_INT) {
        this._serverIterator = 0;
    }

    return this._servers[this._serverIterator++ % this._servers.length];
};


Graylog.prototype._getClient = function () {
    if (!this._client) {
        this._client = dgram.createSocket('udp4');

		this._client.unref();

		var that = this;

        this._client.on('error', function (error) {
			// When a callback is passed to client.send(), this event does not fire.
            that.emit('error', error);
        });
    }

    return this._client;
};


Graylog.prototype.destroy = function () {
	this._sendQueue = null;
	this._deflateQueue = null;
	this._headerPool = [];
	this._isDeflating = false;
	this._isSending = false;

    if (this._client) {
        this._client.close();
        this._client.removeAllListeners();
        this._client = null;
	}
};

Graylog.prototype.emergency = function (short, full, fields, timestamp) {
    this._log(this.level.EMERG, short, full, fields, timestamp);
};

Graylog.prototype.alert = function (short, full, fields, timestamp) {
    this._log(this.level.ALERT, short, full, fields, timestamp);
};

Graylog.prototype.critical = function (short, full, fields, timestamp) {
    this._log(this.level.CRIT, short, full, fields, timestamp);
};

Graylog.prototype.error = function (short, full, fields, timestamp) {
    this._log(this.level.ERROR, short, full, fields, timestamp);
};

Graylog.prototype.warning = function (short, full, fields, timestamp) {
    this._log(this.level.WARNING, short, full, fields, timestamp);
};

Graylog.prototype.notice = function (short, full, fields, timestamp) {
    this._log(this.level.NOTICE, short, full, fields, timestamp);
};

Graylog.prototype.info = function (short, full, fields, timestamp) {
    this._log(this.level.INFO, short, full, fields, timestamp);
};

Graylog.prototype.debug = function (short, full, fields, timestamp) {
    this._log(this.level.DEBUG, short, full, fields, timestamp);
};

Graylog.prototype.log = Graylog.prototype.info;
Graylog.prototype.warn = Graylog.prototype.warning;


Graylog.prototype._serialize = function (level, short, full, fields, timestamp) {
    var message = {
        version: '1.0',
        timestamp: parseInt((timestamp || Date.now()) / 1000, 10),
        host: this._hostname,
        facility: this._facility,
        level: level,
        short_message: short === null ? undefined : short,
        full_message: full === null ? undefined : full
    };

    // We insert additional fields

	if (fields) {
		for (var field in fields) {
	        if (field === 'id') {
	            // http://docs.graylog.org/en/2.1/pages/gelf.html
	            message.__id = fields.id;
	        } else {
	            message['_' + field] = fields[field];
	        }
	    }
	}

    return Buffer.from(JSON.stringify(message), 'utf8');
};


Graylog.prototype._getHeadersFromPool = function (n) {
	for (var i = this._headerPool.length; i < n; i += 1) {
		var header = this._headerPool[i] = Buffer.alloc(12);

		// Set the magic number (bytes 0 and 1)
        header[0] = 30;
        header[1] = 15;

		// Set the chunk sequence number (byte 10)
        header[10] = i;
	}

	return this._headerPool;
};


Graylog.prototype._sendChunked = function (id, message, cb) {
	var maxDataSize = this._bufferSize - 12; // the message part of each chunk is the buffer size - header size
	var chunkCount = Math.ceil(message.length / maxDataSize);

	if (chunkCount > 128) {
		return cb(new Error('Graylog2 message too long: ' + message.length + ' bytes'));
	}

	var client = this._getClient();
	var server = this._getServer();

	var headers = this._getHeadersFromPool(chunkCount);
	var msgOffset = 0;

	for (var i = 0; i < chunkCount; i += 1) {
		var header = headers[i];

		// Set the message id (bytes 2-9)
		id.copy(header, 2);

		// Set the total number of chunks (byte 11)
		header[11] = chunkCount;

		// Slice out the message part
		var data = message.slice(msgOffset, msgOffset + maxDataSize);

		this.emit('chunk', header, data, i, chunkCount, server);

		if (i < chunkCount - 1) {
			client.send([header, data], server.port, server.host);

			msgOffset += maxDataSize;
		} else {
			client.send([header, data], server.port, server.host, cb);
		}
	}
};


var count = 0;

Graylog.prototype._tickDeflate = function () {
	if (this._isDeflating || this._deflateQueue.isEmpty()) {
		return;
	}

	this._isDeflating = true;

	var that = this;
	var msg = this._deflateQueue.getOne();

	function done() {
		that._isDeflating = false;
		that._sendQueue.append(msg);

		that._tickSend();
		that._tickDeflate();
	}

	if (!this._alwaysDeflate && msg.buff.length <= this._bufferSize) {
		process.nextTick(done);
        return;
    }

    zlib.deflate(msg.buff, function (error, compressed) {
        if (error) {
            that.emit('warning', error);
        } else {
			that.compressed += 1;

			if (that._alwaysDeflate || compressed.length < msg.buff.length) {
				msg.buff = compressed;
			}
        }

		done();
    });
};


Graylog.prototype._tickSend = function () {
	if (this._isSending) {
		return;
	}

	if (this._sendQueue.isEmpty()) {
		if (!this._isDeflating) {
			this.emit('drain');
		}
		return;
	}

	this._isSending = true;

	var that = this;
	var msg = this._sendQueue.getOne();

	function done(error) {
		that._isSending = false;

		if (error) {
			that.emit('error', error);
		}

		that.sent += 1;
		that._tickSend();
	}

	var buff = msg.buff;
	msg.buff = null; // help GC a bit

    if (buff.length <= this._bufferSize) {
		// No need to chunk this message

		var client = this._getClient();
		var server = this._getServer();

		this.emit('message', buff, server);

		client.send(buff, 0, buff.length, server.port, server.host, done);
		return;
    }

    // Generate a random ID (buffer)
    crypto.randomBytes(8, function (error, id) {
        if (error) {
            return done(error);
        }

		that._sendChunked(id, buff, done);
    });
};


Graylog.prototype._log = function log(level, short, full, fields, timestamp) {
    if (!this._sendQueue) {
		// destroyed
		this.emit('warning', new Error('Trying to send on a closed client'));
        return;
    }

    var message = {
		buff: this._serialize(level, short, full, fields, timestamp),
		next: null
	};

	if (!this._neverDeflate) {
		this._deflateQueue.append(message);
		this._tickDeflate();
	} else {
		this._sendQueue.append(message);
		this._tickSend();
	}
};


Graylog.prototype.close = function (cb) {
	if (!cb) {
		cb = function () {};
	}

    if (!this._isSending && !this._isDeflating) {
        this.destroy();
        return cb();
    }

    var that = this;

    this.once('drain', function () {
        that.destroy();
		cb();
    });
};


module.exports = Graylog;
